Use file with GPG keys for diagnostics script and other improvements

anders-elastisys opened the pull request:
<!-- Choose your PR title carefully as it will be used as the entry in the changelog! -->

> [!warning]
> **This is a public repository, ensure not to disclose:**
>
> - [x] personal data beyond what is necessary for interacting with this pull request, nor
> - [x] business confidential information, such as customer names.

### What kind of PR is this?

**Required**: Mark one of the following that is applicable:

- [ ] kind/feature       <!-- This PR adds a new feature -->
- [x] kind/improvement   <!-- This PR changes an existing feature -->
- [ ] kind/deprecation   <!-- This PR removes an existing feature -->
- [ ] kind/documentation <!-- This PR contains documentation -->
- [ ] kind/clean-up      <!-- This PR cleans up technical debt -->
- [ ] kind/bug           <!-- This PR fixes a bug -->
- [ ] kind/other         <!-- This PR does something else -->

*Optional*: Mark one or more of the following that are applicable:

> [!important]
> Breaking changes should be marked `kind/admin-change` or `kind/dev-change` depending on type
> Critical security fixes should be marked with `kind/security`

- [x] kind/admin-change <!-- This PR introduces an admin facing change, add "Platform Administrator notice" section -->
- [ ] kind/dev-change   <!-- This PR introduces a dev facing change, add "Application Developer notice" section -->
- [ ] kind/security     <!-- This PR introduces a critical security fix, add "Security notice" section -->
- [ ] [kind/adr]()      <!-- This PR implements an ADR, add the link -->

<!-- Uncomment the additional sections that applies. -->

<!-- Additional information to be added in the release notes
### Release notes
...
-->

### Platform Administrator notice

The diagnostics script now assumes to retrieve GPG keys from a file `$CK8S_CONFIG_PATH/diagnostics_receiver.gpg` if the `CK8S_PGP_FP` environment variable is not set.

<!-- Add additional information with kind/dev-change
### Application Developer notice
...
-->

<!-- Add additional information with kind/security
### Security notice
...
-->

### What does this PR do / why do we need this PR?

Want to simplify the usage of the diagnostics script, the script will now by default import GPG keys from a file `$CK8S_CONFIG_PATH/diagnostics_receiver.gpg` and use their fingerprints if `CK8S_PGP_FP` is not already set.

- Follow up of https://github.com/elastisys/ck8s-issue-tracker/issues/258

#### Information to reviewers

I am planning to update the public docs to reflect this change.

<!--
Any additional information reviews should know.

How to run / how to test.

Include screenshots if applicable to help explain these changes.
--->

#### Checklist

<!-- This section is not added to the changelog or release notes, it is to help you as a contributor and reviewers. -->

- [x] Proper commit message prefix on all commits
  <!-- Example of commit message prefixes:
  - all: changes to multiple areas
  - apps: changes to applications running in all clusters
  - apps sc: changes to applications running in the service cluster
  - apps wc: changes to applications running in the workload cluster
  - bin: changes to management binaries or scripts
  - config: changes to configuration
  - docs: changes to documentation
  - release: release related
  - scripts: changes to other scripts
  - tests: changes to tests
  -->
- Change checks:
  - [ ] The change is transparent
  - [ ] The change is disruptive
  - [ ] The change requires no migration steps
  - [ ] The change requires migration steps
  - [ ] The change upgrades CRDs
  - [ ] The change updates the config *and* the schema
- Metrics checks:
  - [ ] The metrics are still exposed and present in Grafana after the change
  - [ ] The metrics names didn't change (Grafana dashboards and Prometheus alerts are not affected)
  - [ ] The metrics names did change (Grafana dashboards and Prometheus alerts were fixed)
- Logs checks:
  - [ ] The logs do not show any errors after the change
- Pod Security Policy checks:
  - [ ] Any changed pod is covered by Pod Security Admission
  - [ ] Any changed pod is covered by Gatekeeper Pod Security Policies
  - [ ] The change does not cause any pods to be blocked by Pod Security Admission or Policies
- Network Policy checks:
  - [ ] Any changed pod is covered by Network Policies
  - [ ] The change does not cause any dropped packets in the `NetworkPolicy Dashboard`
- Audit checks:
  - [ ] The change does not cause any unnecessary Kubernetes audit events
  - [ ] The change requires changes to Kubernetes audit policy
- Falco checks:
  - [ ] The change does not cause any alerts to be generated by Falco
- Bug checks:
  - [ ] The bug fix is covered by regression tests

cristiklein commented:
@anders-elastisys I guess you wanted some early feedback on your proposal.

Overall, I think this is a good idea. It would be great if we could bring the script to the point where the on-call person only needs to type:

```
./bin/ck8s diagnostics sc
./bin/ck8s diagnostics wc
```

The expression [convention over configuration](https://en.wikipedia.org/wiki/Convention_over_configuration) comes to my mind. The on-call person should be provided with "sensible defaults".

anders-elastisys commented:
> @anders-elastisys I guess you wanted some early feedback on your proposal.
> 
> Overall, I think this is a good idea. It would be great if we could bring the script to the point where the on-call person only needs to type:
> 
> ```
> ./bin/ck8s diagnostics sc
> ./bin/ck8s diagnostics wc
> ```
> 
> The expression [convention over configuration](https://en.wikipedia.org/wiki/Convention_over_configuration) comes to my mind. The on-call person should be provided with "sensible defaults".

Maybe https://github.com/elastisys/compliantkubernetes-apps/pull/2250/commits/13414e179fe9f740586312bc50cf4648f1fdf4a1 is more in line with what we want? I removed the sops-config file flag and instead added so that by default the script will look for the file `${CK8S_CONFIG_PATH}/diagnostics_receiver.gpg` containing GPG keys that are then imported to the operators keys and used by SOPS later in the script. It is still possible to set `CK8S_PGP_FP` manually, but I removed the support for using `.sops.yaml` file as I deemed it not necessary for what we want this command to achieve. Let me know what you think.

cristiklein reviewed (approved):
I like the transparency of the error message and that we provide the platform administrator with sane defaults.

salehsedghpour commented on bin/diagnostics.bash:
```diff
@@ -42,37 +47,54 @@ while [ "${#}" -gt 0 ] ; do
+    echo -e "\t\${CK8S_CONFIG_PATH}/diagnostics_receiver.gpg\n" 1>&2
 
-    echo -e "\tCK8S_PGP_FP=<fingerprint provided during onboarding> ./bin/ck8s diagnostics [sc|wc]\n" 1>&2
+    echo -e "\tIf in doubt, contact support@elastisys.com\n" 1>&2
```
Shouldn't our customer contact `sme-support@elastisys.com`?

anders-elastisys replied:
Fixed it now in https://github.com/elastisys/compliantkubernetes-apps/pull/2250/commits/3ad844b283db5cade0dfd3b2e7fff84e2884a792

simonklb commented on bin/diagnostics.bash:
```diff
@@ -42,37 +47,54 @@ while [ "${#}" -gt 0 ] ; do
-    echo -e "\tEncrypting using the fingerprints: $fingerprints." 1>&2
-    echo -e "\tIf you want to send diagnostic data to Elastisys, make sure to do:\n" 1>&2
+    echo -e "\tIf you want to send diagnostic data to Elastisys, make sure to store GPG keys" 1>&2
+    echo -e "\tretrieved during onboarding in a file named:\n" 1>&2
```
This is not appropriate for this repository since it is open source. We should not assume that Elastisys is owning the platform.

anders-elastisys replied:
The notice was added as when the script was first created, see the discussion [here](https://github.com/elastisys/compliantkubernetes-apps/pull/2043#discussion_r1523258042). Maybe @cristiklein can chip in regarding this?

cristiklein replied:
I agree with @simonklb that this is one of the few instance when we talk with "company voice" as opposed to talking with "project voice". As far as I saw from other project, it is okay to advertise support for an open source project. Hence, I propose the following:

```
If you are an Elastisys self-managed customer, you can send diagnostic data to Elastisys.
Make sure to store GPG keys retrieved during onboarding in a file named:
```

simonklb commented on bin/diagnostics.bash:
```diff
@@ -42,37 +47,54 @@ while [ "${#}" -gt 0 ] ; do
+    echo -e "\t\${CK8S_CONFIG_PATH}/diagnostics_receiver.gpg\n" 1>&2
 
-    echo -e "\tCK8S_PGP_FP=<fingerprint provided during onboarding> ./bin/ck8s diagnostics [sc|wc]\n" 1>&2
+    echo -e "\tIf in doubt, contact support@elastisys.com\n" 1>&2
```
This is not appropriate for this repository since it is open source. We should not assume that Elastisys is owning the platform.

simonklb commented on bin/diagnostics.bash:
```diff
@@ -400,8 +420,18 @@ run_diagnostics_query_metric() {
+
 config_load "${cluster}"
+log_info "Running diagnostics..."
+export CK8S_AUTO_APPROVE=true
```
Is there a reason why this is hardcoded instead of up to the user to set?

anders-elastisys replied:
Yes, otherwise the validation will trigger a prompt that is not visible when running the script since we are piping all output to a file in CK8S_CONFIG_PATH. The output from the validation will still be added to the diagnostics file, but with this it won't be seen as the script gets "stuck".

anders-elastisys replied:
And since we already do a validation before running diagnostics, the user will already know if there is something deviating in their config, and hopefully fix it or at least be aware of it.

simonklb replied:
Makes sense. It's important to note that `CK8S_AUTO_APPROVE` doesn't just automatically approve validation failure prompts but _all_ prompts. So be careful with what you run as part of the diagnostics collection. :smile:

anders-elastisys replied:
Good point, I believe currently it is only the validation that has a prompt, but good to keep in mind. 
We should probably improve it somehow in the future, like better indication that the script is actually running and doing stuff, as it can take quite a while to run and sometimes it is unclear if it is running or stuck.

simonklb reviewed (approved):
LGTM! :)
